### ISSUE TITLE
View building should not be considered complete if there are pending view hints

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#19874]
This is a theoretical bug suggested by @bhalevy in https://github.com/scylladb/scylladb/issues/18826#issuecomment-2249577646 - I have not yet proved it actually exists. The theory is the following:

Supposedly (this assumption needs to be verified!), while view-building is in progress on node B, if some view replica C is down, Scylla may write view hints and proceed with the view building. It is possible that eventually the view building will officially end - the user will be able to read from the view build progress table that it has ended - while some of these view hints have still not been replayed and have not reached the view replica. The same problem can happen on more than one view replica - e.g., we can have two view replicas down, and the view updates will not reach them yet. When these nodes come back up, there will be some time when they will be responsive to reads, but will not contain the missing data.

We should compare this to how ordinary _writes_ handle hints. Writes do **not** include hints in CL (except CL=ANY). A write will only be successful if it really got an ack from CL replicas. So when a write succeeded, the caller knows the data really reached CL replicas - not hints. Similarly, when a view build succeeds, we should know it really reached all replicas, not hints.

One counter-argument that could be raised: Imagine that one node is down for a not-short period (e.g., one hour). Should all view builds not finish during that period? This could hurt _availability_ of the system. Should a user not be able to add views while one node in a large cluster is temporarily down?

**Comment by bhalevy** [comment]
1. Maybe we should add a result status column to the view building status table so when done, the user can check if there were any errors when building the view and act upon it (wait on hints, repair the view table, etc.)

2. I'm not sure comparing view building to tegilar writes is fair, since the former is a long, system-wide process while the latter is momentarly. Currently we don't allow bootstrap when nodes are down.  We could with RBNO by rebuilding ranges that need to be migrated from the DOWN node to the new node.  Maybe comparing to a topology change makes more sense.

**Comment by nyh** [comment]
> Maybe we should add a result status column to the view building status table so when done, the user can check if there were any errors when building the view and act upon it (wait on hints, repair the view table, etc.)

Note that the view building status table is identical to Cassandra and might be considered part of our API (see ongoing discussion https://github.com/scylladb/scylladb/pull/19745 on exactly this issue). I guess we can add additional information somehow. I'm not sure if the users will really notice it.

**Comment by wmitros** [comment]
If I'm not mistaken, we don't actually allow hints during view building - in `view_update_generator::populate_views` we have:
https://github.com/scylladb/scylladb/blob/f3f956841fc4fb720626de2d036478d64c76dd3a/db/view/view_update_generator.cc#L362-L363
The `service::allow_hints::no` here translates to writing view updates with `cl=ONE`:
https://github.com/scylladb/scylladb/blob/f3f956841fc4fb720626de2d036478d64c76dd3a/service/storage_proxy.cc#L4050
And unless we have `cl=ANY` we don't write hints unless we reach `cl_achieved`, which with `cl=ONE` is basically a successful write.

So the question remains if we're actually okay with the _availability_ reduction that this causes - but I haven't seen any issues posted with this problem, so it's probably also okay

**Comment by piodul** [comment]
> So the question remains if we're actually okay with the availability reduction that this causes - but I haven't seen any issues posted with this problem, so it's probably also okay

As I understand this the issue was opened to track a hypothetical bug which was proved not to exist. I'll go ahead and close.

If there is a request to implement the possibility to build views with better availability but potentially lower consistency then I think it should be a separate issue and a separate discussion.
